cmd/osm-controller/main.go

```diff
@@ -102,40 +130,73 @@ func main() {
 		klog.Fatalf("invalid cluster dns specified: %v", err)
 	}
 
-	mgr, err := manager.New(config.GetConfigOrDie(), manager.Options{})
+	logger, err := zap.NewProduction()
 	if err != nil {
-		klog.Error(err, "could not create manager")
-		os.Exit(1)
-	}
-	if err = v1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
-		klog.Fatal(err)
-	}
-	// because we watch MachineDeployments
-	if err = clusterv1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
 		klog.Fatal(err)
 	}
-	logger, err := zap.NewProduction()
+	log := logger.Sugar()
+
+	// Create manager with client against in-cluster config
+	mgr, err := createManager(opt)
 	if err != nil {
-		klog.Fatal(err)
+		klog.Fatalf("failed to create runtime manager: %v", err)
 	}
 
-	log := logger.Sugar()
+	// Start with assuming that current cluster will be used as worker cluster
+	workerClient := mgr.GetClient()
 
-	if err := osp.Add(mgr, log, opt.namespace, opt.workerCount); err != nil {
-		klog.Fatal(err)
+	// Handling for worker cluster
+	if opt.workerClusterKubeconfig != "" {
+		workerClusterConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: opt.workerClusterKubeconfig},
+			&clientcmd.ConfigOverrides{}).ClientConfig()
+		if err != nil {
+			klog.Fatal(err)
+		}
+
+		workerClusterMgr, err := manager.New(workerClusterConfig, manager.Options{
+			LeaderElection:          true,
+			LeaderElectionID:        "operating-system-manager-worker-manager",
+			LeaderElectionNamespace: defaultLeaderElectionNamespace,
+			HealthProbeBindAddress:  opt.workerHealthProbeAddress,
+			MetricsBindAddress:      opt.workerMetricsAddress,
+			Port:                    9444,
+		})
+		if err != nil {
+			klog.Fatal(err)
+		}
+
+		// "-worker-cluster-kubeconfig" was not empty and a valid kubeconfig was provided,
+		// point workerClient to the external cluster
+		workerClient = workerClusterMgr.GetClient()
+
+		// Use workerClusterKubeconfig since the machines will exist on that cluster
+		opt.kubeconfig = opt.workerClusterKubeconfig
+
+		if err := mgr.Add(workerClusterMgr); err != nil {
```

Comment:
@moadqassem  this doesn't seem to work i.e. the events against `MachineDeployment` on worker cluster are not polled.

Reply:
You don't need the osp controller to anything about the worker node cluster. Please take a look how the controller is implemented